### PR TITLE
Rootio : MarshalROOT

### PR DESCRIPTION
**Pull request by maloft**
I tried to recreate the same test as in rbuffer_test.go, but didn't go that well.
Please, need ur review :)

### CODE REVIEW

**Comment by maloft** [comment]
Hey,

hope you arrived well !

Sorry for bothering you this early, but The cherry-pick command does'nt
work, it says :
 Commit fae23c2eaf1e64a9af07b152d597b3dcf7292d10 is a merge but no -m
option was given.
what is it about ?

Le ven. 6 juil. 2018 à 10:26, Sebastien Binet <notifications@github.com> a
écrit :

**Comment by maloft** [comment]
hey, me again,

I got some help so i figured out what the problem is. Still trying to fix
it.
-a

**Comment by maloft** [comment]
Hello,
Please, if possible Can I have some kind of first push for this MarshalROOT
task? I find it a little harder than the WBuffer one.
Thanks
-a

**Comment by sbinet** [comment]
have a look at 0455b23

**Comment by sbinet** [comment]
please gofmt your files.

**Comment by sbinet** [comment]
please rebase on top of `4a0ea95` (that fixes the API breakage from gonum/optimize.)

**Comment by sbinet** [comment]
and once that's done, please squash the commit that adds the `TObjString` test together with the one that implements `TObjString.ROOTMarshaler`.

File: rootio/wbuffer_test.go [hunk 1]
```diff
@@ -39,6 +39,17 @@ func TestWBuffer_WriteString(t *testing.T) {
 	}
 }
 
+func TestWBuffer_WriteObjectAy(t *testing.T) {
```
Comment (sbinet) [inline rootio/wbuffer_test.go]: `s/WriteObjectAy/WriteObjectAny/`

File: rootio/wbuffer.go [hunk 2]
```diff
@@ -94,20 +94,22 @@ func (w *WBuffer) WriteObjectAny(obj Object) error {
...
 }
 
 func (w *WBuffer) WriteClass(obj Object) error {
 	if w.err != nil {
 		return w.err
 	}
-	//	class := obj.Class()
+	class := obj.Class()
+	w.WriteString(class)
+	w.w.c += len(class)
```
Comment (sbinet) [inline rootio/wbuffer.go]: not sure that's really what's needed. (the failing test seems to agree with me :P)

File: rootio/wbuffer.go [hunk 3]
```diff
@@ -94,20 +94,22 @@ func (w *WBuffer) WriteObjectAny(obj Object) error {
...
-	//	w.WriteU32(0) // placeholder for bytecount.
-	//
-	//	mapsize := len(w.refs)
-	//	w.WriteClass(obj)
+	//pos := w.Pos()
+	w.WriteU32(0) // placeholder for bytecount.
 
+	mapsize := len(w.refs)
+	w.WriteClass(obj)
+	w.w.c += mapsize
```
Comment (sbinet) [inline rootio/wbuffer.go]: not sure that's really what's needed. (the failing test seems to agree with me :P)

**Comment by maloft** [comment]
The last commit is only a try. Possible help needed :)

File: rootio/utils.go [hunk 4]
```diff
@@ -53,7 +53,7 @@ func datime2time(d uint32) time.Time {
 }
 
 // time2datime converts a time.Time into a uint32
-func time2datime(t time.Time) uint32 {
+func time2datime(t time.Time) uint32{
```
Comment (sbinet) [inline rootio/utils.go]: please gofmt this.

**Comment by sbinet** [comment]
still not properly `gofmt`'d:
- https://travis-ci.org/go-hep/hep/jobs/408545459

also, it seems your `rootio-write` branch has been merged with `origin/master`:
```
* 11c0d83 (origin/csvutil-coverage) csvutil: add tests for all supported types
| *   d321e8d (maloft/rootio-write) rootio: implementing WriteObjectAny (third try)
| |\  
| * | dc90477 rootio: implementing WriteObjectAny (second shot)
| * | 1275549 rootio: trying to implement WriteObjectAny + test
| * | b51b0b3 rootio: implement ROOTMarshaler for TKey
| * | 440c931 rootio: implement ROOTMarshaler for TObjstring
| * | 8051dee rootio: implement ROOTMarshaler for TList (first stab)
| * | c1cc058 rootio: first stab at wbuffer WriteObjectAny and WriteClass
| * | 478f9f1 rootio: implement ROOTMarshaler for TH1
| * | 66149fc rootio: implement ROOTMarshaler for TAxis
| * | e63cf77 rootio: implement ROOTMarshaler for TH1{F,D,I}
| * | c69cea1 rootio: implement ROOTMarshaler for TAttMarker
| * | 78f30dd rootio: implement ROOTMarshaler for TAttLine
| * | 399cd06 rootio: implement ROOTMarshaler for TAttFill
| * | a98535a rootio: implement ROOTMarshaler for TAttAxis
|/ /  
| | * 53d9408 (origin/rootio-rw-th1) rootio: implement ROOTMarshaler for TH1{F,D,I}
| | * d85b21a rootio: implement ROOTMarshaler for TH1
| | * 171aee5 rootio: implement ROOTMarshaler for TAxis
| | * b4c4086 rootio: implement ROOTMarshaler for TAttMarker
| | * fb4236e rootio: implement ROOTMarshaler for TAttLine
| | * 436a820 rootio: implement ROOTMarshaler for TAttFill
| | * d64d7cf rootio: implement ROOTMarshaler for TAttAxis
| | * eb73f93 rootio: implement ROOTMarshaler for TList (first stab)
| | * cdb19b4 rootio: first stab at wbuffer WriteObjectAny and WriteClass
| |/  
| * 3e3501b (origin/master, origin/HEAD) rootio: implement time2datime
| * a86eb7d rootio: add test for decodeNameCycle
|/  
* 8e48554 all: remove extraneous space in LICENSE blurb
* 9e10951 rootio: implement ROOTMarshaler for TArrays
* 683ce96 fit: update for new gonum/optimize.Minimize API
```
please keep the `git` history linear: rebase on top of the latest `origin/master`.

**Comment by maloft** [comment]
Please when possible, need help with WriteObjectAny.
Thanks

Le jeu. 26 juil. 2018 à 17:59, Sebastien Binet <notifications@github.com> a
écrit :

**Comment by sbinet** [comment]
`WriteObjectAny` is now in https://github.com/go-hep/hep/pull/314.
(about to be merged, once the final CI run is green.)

please rebase on top of it so we can merge the `TObjString` stanza.

**Comment by sbinet** [comment]
this PR is mostly irrelevant now (the only interesting bits are the `TKey` marshaling implementation, which should be easily extractable in their own PR.)

closing.

**Pull request closed without merging**